wrappers/android/zxingcpp/src/main/java/com/zxingcpp/ZXingCpp.kt

```diff
@@ -0,0 +1,152 @@
+/*
+* Copyright 2021 Axel Waggershauser
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+package com.zxingcpp
+
+import android.graphics.Bitmap
+import android.graphics.ImageFormat
+import android.graphics.Point
+import android.graphics.Rect
+import android.os.Build
+import androidx.camera.core.ImageProxy
+import java.nio.ByteBuffer
+
+public object ZXingCpp {
+	private val supportedYUVFormats: List<Int> =
+		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+			listOf(ImageFormat.YUV_420_888, ImageFormat.YUV_422_888, ImageFormat.YUV_444_888)
+		} else {
+			listOf(ImageFormat.YUV_420_888)
+		}
+
+	init {
+		System.loadLibrary("zxing_android")
+	}
+
+	// Enumerates barcode formats known to this package.
+	// Note that this has to be kept synchronized with native (C++/JNI) side.
+	public enum class Format {
+		NONE, AZTEC, CODABAR, CODE_39, CODE_93, CODE_128, DATA_BAR, DATA_BAR_EXPANDED,
+		DATA_MATRIX, EAN_8, EAN_13, ITF, MAXICODE, PDF_417, QR_CODE, MICRO_QR_CODE, UPC_A, UPC_E
+	}
+
+	public enum class ContentType {
+		TEXT, BINARY, MIXED, GS1, ISO15434, UNKNOWN_ECI
+	}
+
+	public enum class Binarizer {
+		LOCAL_AVERAGE, GLOBAL_HISTOGRAM, FIXED_THRESHOLD, BOOL_CAST
+	}
+
+	public enum class EanAddOnSymbol {
+		IGNORE, READ, REQUIRE
+	}
+
+	public enum class TextMode {
+		PLAIN, ECI, HRI, HEX, ESCAPED
+	}
+
+	public data class DecodeHints(
+		var formats: Set<Format> = setOf(),
+		var tryHarder: Boolean = false,
+		var tryRotate: Boolean = false,
+		var tryInvert: Boolean = false,
+		var tryDownscale: Boolean = false,
+		var isPure: Boolean = false,
+		var tryCode39ExtendedMode: Boolean = false,
+		var validateCode39CheckSum: Boolean = false,
+		var validateITFCheckSum: Boolean = false,
+		var returnCodabarStartEnd: Boolean = false,
+		var returnErrors: Boolean = false,
+		var downscaleFactor: Int = 3,
+		var eanAddOnSymbol: EanAddOnSymbol = EanAddOnSymbol.IGNORE,
+		var binarizer: Binarizer = Binarizer.LOCAL_AVERAGE,
+		var textMode: TextMode = TextMode.HRI,
+		var minLineCount: Int = 2,
+		var maxNumberOfSymbols: Int = 0xff,
+		var downscaleThreshold: Int = 500
+	)
+
+	public data class Position(
+		val topLeft: Point,
+		val topRight: Point,
+		val bottomLeft: Point,
+		val bottomRight: Point,
+		val orientation: Double
+	)
+
+	public data class Result(
```

Comment:
There are some inconsistencies that I'd like to address:
 * You added `returnErrors` to `DecodeHints` but there is no corresponding `error` member here.
 * Some members are 'nullable', others are not. Whether or not this result describes a valid or erroneous symbol we always have a format, position, orientation and time (and arguably a sequence size and index). Everything else is optional / depending on the validity of the result. I understand that this optionality is canonically described with the `?` in Kotlin. Not sure whether a `Boolean?` would people make rise their eyebrows?

Reply:
> You added returnErrors to DecodeHints but there is no corresponding error member here.

Ah, my bad, of course. I've just added a commit to add an `error` member to `Result`. For the sake of simplicity I just return the `msg` string, but of course could add an Error structure with the corresponding type, if you wish so.

> Some members are 'nullable', others are not…

Well, I didn't change the existing members (to keep the changes as minimal as possible), and `time` and `postion` were optional before. But that really doesn't make much sense because they're always set on the [C++ side](https://github.com/zxing-cpp/zxing-cpp/pull/659/files#diff-b0fe4e53e8e58acbf0322650c3f08f1237c7d4a80738eb759a5293426224046fR199).

Now, that is also true all other members, which is why I have no optionals in the Result of my [wrapper](https://github.com/markusfisch/zxing-cpp/blob/master/wrappers/aar/zxingcpp/src/main/java/de/markusfisch/android/zxingcpp/ZxingCpp.kt#L99) (except of `gtin`, which we don't want here). I think this is the best match of the native Result, where `Result.text()` always returns a `std::string` object, and never `NULL`. So I thought the Kotlin/JVM Result should also behave that way and have an empty `String` instead of an optional `String?`.

Also, we can't really make something like `readerInit` a `Boolean?` because the native `Result.readerInit()` is just a `bool`. So `Result.readerInit` on the JVM will always be non-null, too.

So my suggestion would be to remove all optionals in Result 😉

Reply:
If we can agree that `text : String?` is the correct model, then `sequenceId` should also be optional.

Reply:
There's now a proper `Error` struct, `sequenceId` is nullable, and `Result.text` will now be `null` when the `Result` is not valid (as are the other nullable members of `Result`).

> But in Kotlin I got the impression that having a String? immediately communicates to the user that this thing can be nonexisting and there are language features to nicely deal with that.

Yes, exactly.

> readerInit could be set to null based on Result::isValid() is we really wanted to. I'm not convinced that this is required, just saying ;).

I skipped that for now 😉 If you want, I can change it to `Boolean?`, too. Although it might make more sense to add a `valid` member instead of making all members optional.